### ISSUE TITLE
Optimizations for clients & servers

### ISSUE THREAD

**Issue by Raynos** [uber/tchannel-node#237]
I'm working on a fat heavy branch and I'm keeping notes of optimizations that we should backport everywhere

 - [ ] use `this` inside constructors, v8 literally pattern detects for `this.foo = x` and NOT `self.foo =x`
 - [ ] use bucket based `OutPending()` data structure instead of naive dictionary

cc @jcorbin @rf @kriskowal

**Comment by rf** [comment]
Related to the first point -- the `instanceof` check really does happen every time and does have a cost.

**Comment by Raynos** [comment]
Yep; we need to avoid the `instanceof` pattern in any constructors that are not public interface or any constructors that are O(n) requests or O(n) sockets

We may also need to rethink the `instanceof` pattern for public interfaces and maybe favor a node-core-style `createX()` function.
